libs/wai-utilities/src/Network/Wai/Utilities/Server.hs

```diff
@@ -185,21 +168,43 @@ measureRequests m rtree = withPathTemplate rtree $ \p ->
       requestCounter m p . duration 30 12 m p
 {-# INLINEABLE measureRequests #-}
 
+-- | Run both 'catchErrorsException', 'catchErrorsResponse'.
+catchErrors  :: Logger -> OnErrorMetrics -> Middleware
+catchErrors l m = catchErrorsException l m . catchErrorsResponse l
+{-# INLINEABLE catchErrors #-}
+
 -- | Create a middleware that catches exceptions and turns
 -- them into appropriate 'Error' responses, thereby logging
 -- as well as counting server errors (i.e. exceptions that
 -- yield 5xx responses).
-catchErrors :: Logger -> Metrics -> Middleware
-catchErrors l m app req k =
+--
+-- This does not log any 'Response' values with error status.
+-- See 'catchErrors'.
+catchErrorsException :: Logger -> OnErrorMetrics -> Middleware
+catchErrorsException l m app req k =
     app req k `catch` errorResponse
   where
     errorResponse ex = do
         er <- runHandlers ex errorHandlers
         when (statusCode (Error.code er) >= 500) $
-            logIO l Log.Error (Just req) (oneline <$> show ex)
+            logIO l Log.Error (Just req) (show ex)
         onError l m req k er
-    oneline c = if isSpace c then ' ' else c
-{-# INLINEABLE catchErrors #-}
+{-# INLINEABLE catchErrorsException #-}
+
+-- | Catch status >= 400 even if it does not come in the form of an exception, but as a
+-- response, and log a warning that this really shouldn't happen, and a typed exception should
+-- be thrown instead.
+catchErrorsResponse :: Logger -> Middleware
+catchErrorsResponse l app req k =
+    app req $ \resp -> do
+        when (statusCode (responseStatus resp) >= 400) $ do
+            let errinfo = Error (responseStatus resp) (cs . statusMessage $ responseStatus resp) "N/A"
```

Comment:
Why are we clobbering the responses message with an 'N/A' here? Is it so we don't leak server errors? (this shouldn't ever happen anyways; but just curious why we do it).

Reply:
what else do you want to put in there?  we don't have the streaming body of the response (at least it's hard to get to that), and i just couldn't think of anything else that would help there.

Reply:
Hrmm, didn't realize it was so tricky to get the body out; but it seems wrong to just clobber the body entirely; doesn't this imply though that the user is going to get the response body `status: 400; msg: N/A`? That would be very frustrating to me. This should never happen so maybe it's not a big deal; but it would be nice to give something more helpful; even the description of the status code. 

I won't push very hard on this; but it's possible this middleware will be used across all services; some of which may be using responses for many cases, losing the response bodies there would make those problems VERY difficult to debug.

Reply:
if this happens, we have a server error.  we (hopefully) have enough information to fix the server error.  if not, perhaps we can add something from the request, but that's tricky because it may contain passwords (as may the response, btw).

it's not a step backwards.  this doesn't hit the client.  the client will get the original response.  we don't re-throw, this middleware just logs the warning and moves on.

see the commit message in 711318defc9fea16db1a640af3bf6e7efba8b5ab